File: 2023/day6/main.py
input = open("input").read().strip().split("\n")
times = [int(x) for x in input[0].split(":")[1].split()]
distances = [int(x) for x in input[1].split(":")[1].split()]

def p1():
  a = 1;

  for i, time in enumerate(times):
    wins = 0;
    for j in range(time):
      if (time - j) * j > distances[i]:
        wins += 1;
    a *= wins;
  
  return a;

def p2():
  time = int(''.join(str(x) for x in times));
  distance = int(''.join(str(x) for x in distances));
  wins = 0;
  for j in range(time):
    if (time - j) * j > distance:
      wins += 1;

  return wins;

print("Part 1:", p1())
print("Part 2:", p2())

File: 2023/day6/main.go
package main

import (
	"os"
	"fmt"
	"strings"
	"strconv"
)

func stringToInts(strs []string) ([]int, error) {
	ints := make([]int, len(strs))
	for i, str := range strs {
			n, err := strconv.Atoi(str)
			if err != nil {
				return nil, err 
			}
			ints[i] = n
	}
	return ints, nil
}

func parseData() ([]int, []int) {
	content, _ := os.ReadFile("input")
	lines := strings.Split(string(content), "\n")
	times, _ := stringToInts(strings.Fields(strings.Split(lines[0], ":")[1]))
	distances, _ := stringToInts(strings.Fields(strings.Split(lines[1], ":")[1]))
	return times, distances
}

func part1() (int) {
	times, distances := parseData()

	a := 1;
	for i, t := range times {
		wins := 0
		for j := 1; j < t; j++ {
			d := (t - j) * j
			if d > distances[i] {
				wins++
			}
		}
		a *= wins
	}

	return a
}

func part2() (int) {
	times, distances := parseData()
	time := ""
	for _, t := range times {
		time += strconv.Itoa(t)
	}
	distance := ""
	for _, d := range distances {
		distance += strconv.Itoa(d)
	}

	t, _ := strconv.Atoi(time)
	d, _ := strconv.Atoi(distance)

	wins := 0
	for i := 1; i < t; i++ {
		dd := (t - i) * i
		if dd > d {
			wins++
		}
	}

	return wins
}

func main() {
	p1 := part1()
	p2 := part2()

	fmt.Println("Part 1:", p1)
	fmt.Println("Part 2:", p2)
}